### PR TITLE
PR includes buildkite pipeline changes to build electron binaries

### PR DESCRIPTION
**Pull request by twitharshil**

### CODE REVIEW

File: .buildkite/build-and-test-pipeline.js [hunk 1]
```diff
@@ -36,44 +36,11 @@ function buildStepForNix (platform) {
...
-    agents: [
-      `os=${platform}`,
-      'queue=electron-build-v7.2'
-    ]
-  };
-}
-
 function generateBuildPipeline () {
-  // Do not run the pipeline if a PR has not been raised yet
-  // if (process.env.BUILDKITE_PULL_REQUEST === 'false') {
```
Comment (harryi3t) [inline .buildkite/build-and-test-pipeline.js:65]: why don't we need this?

File: .buildkite/build-and-test-pipeline.js [hunk 2]
```diff
@@ -36,44 +36,11 @@ function buildStepForNix (platform) {
...
-  //   return [];
-  // }
-
   return [
     buildStepForWindows(),
     buildStepForNix('linux'),
     buildStepForNix('darwin'),
-    // waitStep(),
-    // testStepForWindows(),
```
Comment (harryi3t) [inline .buildkite/build-and-test-pipeline.js:74]: why there are no tests?

Comment (harryi3t) [reply .buildkite/build-and-test-pipeline.js:74]: we need to build and then test too

File: .buildkite/nix/build-and-test.sh [hunk 3]
```diff
@@ -34,6 +33,12 @@ buildAndUpload() {
   echo "--- Swtiching directory <pipeline>/src/electron"
   cd src/electron
 
+  echo "Cleaning .git files before running gclient sync"
+  cd ..
+  rm -rf .git/rebase-apply
```
Comment (harryi3t) [inline .buildkite/nix/build-and-test.sh:41]: add doc to explain this

File: .buildkite/nix/build-and-test.sh [hunk 4]
```diff
@@ -106,21 +111,28 @@ buildAndUpload() {
   ninja -C out/Release electron:electron_mksnapshot_zip
   
 
-  echo "--- Generate type declarationsp (Linux)"
+
```
Comment (harryi3t) [inline .buildkite/nix/build-and-test.sh]: 2 extra lines

File: .buildkite/nix/build-and-test.sh [hunk 5]
```diff
@@ -106,21 +111,28 @@ buildAndUpload() {
...
   if [[ "$platform" == "linux" ]]
+  echo "--- Generate type declarationsp [Linux]"
```
Comment (harryi3t) [inline .buildkite/nix/build-and-test.sh]: typo

File: .buildkite/nix/build-and-upload-release.sh [hunk 6]
```diff
@@ -43,9 +49,9 @@ buildAndUpload() {
   git checkout $BUILDKITE_BRANCH
   git branch --set-upstream-to origin/$BUILDKITE_BRANCH
 
-  echo "git pull"
-  git pull
-  
+  echo "git reset --hard origin"
+  git reset --hard origin/$BUILDKITE_BRANCH
```
Comment (harryi3t) [inline .buildkite/nix/build-and-upload-release.sh:56]: are we doing `git fetch` somewhere? 
without that it won't take latest changes while pushing new commits to the same branch

File: .buildkite/nix/build-and-upload-release.sh [hunk 7]
```diff
@@ -67,8 +73,8 @@ buildAndUpload() {
   echo "--- Electron build"
   ninja -C out/Release electron
 
-  echo "--- Strip Electron binaries (Linux)"
   if [[ "$platform" == "linux" ]]
+  echo "--- Strip Electron binaries (Linux)"
```
Comment (harryi3t) [inline .buildkite/nix/build-and-upload-release.sh]: indentation

File: .buildkite/nix/build-and-upload-release.sh [hunk 8]
```diff
@@ -105,49 +111,51 @@ buildAndUpload() {
   fi
   ninja -C out/Release electron:electron_mksnapshot_zip
   
-
-  echo "--- Generate type declarationsp (Linux)"
   if [[ "$platform" == "linux" ]]
+  echo "--- Generate type declarationsp (Linux)"
```
Comment (harryi3t) [inline .buildkite/nix/build-and-upload-release.sh]: indentation

File: .buildkite/nix/build-and-upload-release.sh [hunk 9]
```diff
@@ -105,49 +111,51 @@ buildAndUpload() {
...
+  buildkite-agent artifact upload "*.sha256sum"
+  cd ../..
+  # Upload SHA files for electron-api.json and electron.ts
+  cd electron
+  buildkite-agent artifact upload "*.sha256sum"
+  cd ..
 }
 
+  
+
```
Comment (harryi3t) [inline .buildkite/nix/build-and-upload-release.sh]: so many extra line breaks

File: .buildkite/release-and-publish.sh [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+upload_SHA_file_and_release() {
+  echo "--- Switching directory src/electron"
+  cd src/electron
+
+  echo "--- npm i"
+  npm i
+
+  echo "--- Creating directory script/dist"
+  mkdir script/dist/ || true
```
Comment (harryi3t) [inline .buildkite/release-and-publish.sh]: is script already present?

Comment (harryi3t) [reply .buildkite/release-and-publish.sh]: why using this directory which is not git ignored and not cleaned up?
can't we use `dist` or `out` and create some folder in that?

File: .buildkite/release-and-publish.sh [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  echo "Releasing"
+  node script/release/release.js --skipVersionCheck
+}
+
+main() {
+  sanity
+
+  echo "Upload SHA file"
+  upload_SHA_file_and_release
+
```
Comment (harryi3t) [inline .buildkite/release-and-publish.sh]: extra line break

File: .buildkite/release-pipeline.yml [hunk 12]
```diff
@@ -19,20 +19,16 @@ steps:
...
-    # agents:
-    #   nix: "true"
-    #   queue: "electron-build"
+  - wait
+  - label: ":github: Release on Github"
+    command:
+      - ".buildkite/release-and-publish.sh"
+    agents:
+      os: "darwin"
+      queue: "electron-build-v7.2"
```
Comment (harryi3t) [inline .buildkite/release-pipeline.yml]: leave a line break at the end of every file

File: .buildkite/windows/build-and-upload-release.bat [hunk 13]
```diff
@@ -35,13 +41,13 @@ CALL cd /D .. || EXIT /b !errorlevel!
 
 ECHO "--- Building electron binaries in Release mode"
 if "%ARCH%" == "ia32" (
-  CALL gn gen out/Release-x32 --args="import(\"//electron/build/args/release.gn\") target_cpu=\"x86\"" || EXIT /b !errorlevel!
```
Comment (harryi3t) [inline .buildkite/windows/build-and-upload-release.bat:38]: why?

File: .buildkite/windows/build-and-upload-release.bat [hunk 14]
```diff
@@ -82,16 +88,22 @@ if "%ARCH%" == "ia32" (
   CALL buildkite-agent artifact upload ffmpeg/ffmpeg.zip || EXIT /b !errorlevel!
 )
 
-ECHO "--- Upload to GitHub release"
+ECHO "--- Switch directory <pipeline>/src"
+CALL cd .. || EXIT /b !errorlevel!
```
Comment (harryi3t) [inline .buildkite/windows/build-and-upload-release.bat]: how can this command fail?

File: script/lib/util.py [hunk 15]
```diff
@@ -252,7 +252,8 @@ def get_electron_exec():
     return '{0}/electron.exe'.format(out_dir)
   elif sys.platform == 'linux':
     return '{0}/electron'.format(out_dir)
-
+  elif sys.platform == 'linux2':
```
Comment (harryi3t) [inline script/lib/util.py]: 😱  what is this?

File: script/lib/util.py [hunk 16]
```diff
@@ -232,7 +232,7 @@ def add_exec_bit(filename):
   os.chmod(filename, os.stat(filename).st_mode | stat.S_IEXEC)
 
 def get_out_dir():
-  out_dir = 'Debug'
+  out_dir = 'Release'
```
Comment (harryi3t) [inline script/lib/util.py:235]: why this change?

File: script/lib/config.py [hunk 17]
```diff
@@ -86,13 +86,16 @@ def is_verbose_mode():
...
-def get_zip_name(name, version, suffix=''):
+def get_zip_name(name, version, suffix='', arch_ia32=''):
   arch = get_target_arch()
   if arch == 'arm':
     arch += 'v7l'
   zip_name = '{0}-{1}-{2}-{3}'.format(name, version, get_platform_key(), arch)
   if suffix:
     zip_name += '-' + suffix
+
+  if(arch_ia32):
```
Comment (harryi3t) [inline script/lib/config.py:97]: how was it working till now?

File: script/release/release.js [hunk 18]
```diff
@@ -286,17 +248,14 @@ async function makeRelease (releaseToValidate) {
...
-
     await createReleaseShasums(draftRelease);
 
     // Fetch latest version of release before verifying
     draftRelease = await getDraftRelease(pkgVersion, true);
     await validateReleaseAssets(draftRelease);
     await publishRelease(draftRelease);
-    console.log(`${pass} SUCCESS!!! Release has been published. Please run ` +
-      `"npm run publish-to-npm" to publish release to npm.`);
+    // console.log(`${pass} SUCCESS!!! Release has been published. Please run ` +
```
Comment (harryi3t) [inline script/release/release.js]: why is this commented out?

**Review by harryi3t — CHANGES_REQUESTED** [review]

File: script/release/release.js (hunk 18 above)
Comment (twitharshil) [reply script/release/release.js:257]: We have decided to not publish electron artifacts to npm using the pipeline. This step will be done manually by us

File: script/lib/config.py (hunk 17 above)
Comment (twitharshil) [reply script/lib/config.py:97]: We are building the windows-32 bit artifact using the cross-platform option provided by the electron to us. Basically building 32-bit artifacts for electron on a 64-bit machine. This code returns us `x64` even if we are building artifacts for a 32-bit machine. 
Electron community CI machines must have a windows-32 bit machine, that's why they have written such scripts. 
For us, in v3.1.8 there was an argument option available to pass the target arch `--target_arch` in the bootstrap.py file which is not available in v7 now. Now we are using gclient and ninja steps

File: script/lib/util.py (hunk 16 above)
Comment (twitharshil) [reply script/lib/util.py:235]: By default distribution binaries output directory is `Debug`. Since in our case, we will not build binaries in Debug mode, I decided to change the default value to `Release`

File: .buildkite/nix/build-and-upload-release.sh (hunk 6 above)
Comment (twitharshil) [reply .buildkite/nix/build-and-upload-release.sh:56]: Yes, we are doing git fetch just before checkout of the branch 
1. git fetch 
2. git checkout branch 
3. git reset --hard origin/branch

File: .buildkite/nix/build-and-test.sh (hunk 5 above)
Comment (twitharshil) [reply .buildkite/nix/build-and-test.sh]: Correcting it

File: .buildkite/nix/build-and-test.sh (hunk 4 above)
Comment (twitharshil) [reply .buildkite/nix/build-and-test.sh]: Addressed

File: .buildkite/build-and-test-pipeline.js (hunk 2 above)
Comment (twitharshil) [reply .buildkite/build-and-test-pipeline.js:74]: Test scripts are failing :(

File: .buildkite/build-and-test-pipeline.js (hunk 1 above)
Comment (twitharshil) [reply .buildkite/build-and-test-pipeline.js:65]: Just wanted to manually start the pipeline build, it will not spoil our pipelines. Since releasing electron binaries happens once a while, it can be considered as good to have

File: .buildkite/nix/build-and-test.sh (hunk 3 above)
Comment (twitharshil) [reply .buildkite/nix/build-and-test.sh:41]: Adding the doc here

File: .buildkite/nix/build-and-upload-release.sh (hunk 7 above)
Comment (twitharshil) [reply .buildkite/nix/build-and-upload-release.sh]: Addressed

File: .buildkite/nix/build-and-upload-release.sh (hunk 9 above)
Comment (twitharshil) [reply .buildkite/nix/build-and-upload-release.sh]: Addressed

File: .buildkite/release-and-publish.sh (hunk 10 above)
Comment (twitharshil) [reply .buildkite/release-and-publish.sh]: Will check this

File: .buildkite/nix/build-and-upload-release.sh (hunk 8 above)
Comment (twitharshil) [reply .buildkite/nix/build-and-upload-release.sh]: Addressed

File: script/release/release.js (hunk 18 above)
Comment (twitharshil) [reply script/release/release.js]: Adding it back

File: .buildkite/windows/build-and-upload-release.bat (hunk 13 above)
Comment (twitharshil) [reply .buildkite/windows/build-and-upload-release.bat:38]: Since we have different steps to build 32 bit and 64-bit artifacts

File: .buildkite/windows/build-and-upload-release.bat (hunk 14 above)
Comment (twitharshil) [reply .buildkite/windows/build-and-upload-release.bat]: Removing 👍

File: script/lib/util.py (hunk 15 above)
Comment (twitharshil) [reply script/lib/util.py]: sys.platform returns `linux2` :p
